boot/assets.go

```diff
@@ -632,3 +632,117 @@ func (o *TrustedAssetsUpdateObserver) Canceled() error {
 	// - reseal with a given state of modeenv
 	return nil
 }
+
+func observeSuccessfulBootWithAssetsForBootloader(m *Modeenv, root string, opts *bootloader.Options) (drop []*trackedAsset, err error) {
+	trustedAssetsMap := &m.CurrentTrustedBootAssets
+	otherTrustedAssetsMap := m.CurrentTrustedRecoveryBootAssets
+	whichBootloader := "run mode"
+	if opts != nil && opts.Role == bootloader.RoleRecovery {
+		trustedAssetsMap = &m.CurrentTrustedRecoveryBootAssets
+		otherTrustedAssetsMap = m.CurrentTrustedBootAssets
+		whichBootloader = "recovery"
+	}
+
+	if len(*trustedAssetsMap) == 0 {
+		// bootloader may have trusted assets, but we are not tracking
+		// any for the boot process
+		return nil, nil
+	}
+
+	// let's find the bootloader first
+	bl, trustedAssets, err := findMaybeTrustedAssetsBootloader(root, opts)
+	if err != nil {
+		return nil, err
+	}
+	if len(trustedAssets) == 0 {
+		// not a trusted assets bootloader, nothing to do
+		return nil, nil
+	}
+
+	cache := newTrustedAssetsCache(dirs.SnapBootAssetsDir)
+	for _, trustedAsset := range trustedAssets {
+		assetName := filepath.Base(trustedAsset)
+
+		// find the hash of the file on disk
+		assetHash, err := cache.fileHash(filepath.Join(root, trustedAsset))
+		if err != nil && !os.IsNotExist(err) {
+			return nil, fmt.Errorf("cannot calculate the digest of existing trusted asset: %v", err)
+		}
+		if assetHash == "" {
+			// no trusted asset on disk, but we booted nonetheless,
+			// at least log something
+			logger.Noticef("system booted without %v bootloader trusted asset %q", whichBootloader, trustedAsset)
+			// given that asset names cannot be reused, clear the
+			// boot assets map for the current bootloader
+			delete(*trustedAssetsMap, assetName)
+			continue
+		}
+
+		// this is what we booted with
+		bootedWith := []string{assetHash}
+		// one of these was expected during boot
+		hashList := (*trustedAssetsMap)[assetName]
+
+		assetFound := false
+		// find out if anything needs to be dropped
+		for _, hash := range hashList {
+			if hash == assetHash {
+				assetFound = true
+				continue
+			}
+			if !isAssetHashTrackedInMap(otherTrustedAssetsMap, assetName, hash) {
```

Comment:
what situation does this handle? I want to say that this is handling assets that were deleted from trustedAssetsMap, but wouldn't we keep the asset in the modeenv during the change even if it was deleted? Or do we delete assets from disk, then reseal without that deleted asset, then reboot and then finally get here via MarkBootSuccessful ?

Reply:
this is a common pattern in the code, it's about the fact that the cache in itself is shared between assets for the run partition and assets for the recovery partition, so we can drop an asset only if it's also not used by the other partition.

Scenario:  initially we have the same grub in seed and boot. We upgrade the one boot, so from the point of view of the hashes of boot we don't need to keep anymore, but actually is still use for recovery so we need to keep in the cache.

Reply:
but to be clear, the situation where we drop assets from the cache is where the asset was deleted on an update, where perhaps grub.cfg was deleted or something like that and should no longer exist on one of the bootloaders correct? 

I understand now that the cache is shared between the recovery and normal bootloader and that the hashes of the same asset name may go out of sync with updates, but afaict this code is not about that situation, this code is more about when an asset should no longer exist at all due to an update?

also perhaps this was unclear due to the line that I commented on being the check, but I'm more concerned about what situation is it when `isAssetHashTrackedInMap(otherTrustedAssetsMap, assetName, hash)` returns false, and thus we enter this `if` block

Reply:
put more succinclty, when do we drop assets by appending to the drop list?

Reply:
reading the unit tests more closely I see that the situation where we drop assets is just in the normal case of changing an asset, where we are dropping the old version of an asset, apologies for the confusion this makes sense now

Reply:
if I have a asset  "loader" that exist both in seed and boot, if before rebooting we had the mappings to hashes:

recovery->"loader"-> a
boot->"loader"-> b,c

and we correctly boot with loader c on the boot partition and reach this code:

hashList is  [b,c]
assetHash is c

so we arrive here with  b but is not in the other map (the other map has just a), so b will be dropped from the cache, it will be dropped in postModeenv:

https://github.com/snapcore/snapd/pull/9273/files#diff-bff1ced30ed4581302ca81b7b53908f9R671

after we have written the new map, which will be:

recovery->"loader"-> a
boot->"loader"-> c

The test it seems we are missing is something like this where we update the same asset the same way in both boot partitions, situation after reboot like:

recovery->"loader"-> oldhash, newhash
boot->"loader"-> oldhash, newhash

oldhash should get dropped (I think the code will do that but I don't think there's a test for that). There are test for the simple cases where !cond  is true or false (TestObserveSuccessfulBootDropCandidateUsedByOtherBootloader)

Reply:
Thanks for the in depth explanation, yes that matches my understanding now, and I think that a test like you describe is a good idea, your patch you pushed LGTM too